### DISCUSSION TITLE
Push database to server by code

### DISCUSSION THREAD

**Discussion by CacaceNicolas** [sqlitebrowser/sqlitebrowser#3537] [Q&A] [answered]
I am currently working on a game using the Godot engine with a plugin called "GodotSQLite". it is a TCG and i was trying to make it possible for players to trade cards online. In order to do this they should all be able to read and write on the same database at the same time. Im using DBHub.io for this, and it worked, but, i can only submit changes to the database using the "Push currently opened database to server" button on the DB Browser. I was wondering if there was a way to do this using a query or some kind of command, so i can update the database from the game itself

TLDR: is there a way to push the currently opened database to server using a query or command from an external app?

Thanks!

**Comment by FriedrichFroebel** [comment]
DBHub.io has its own API: https://api.dbhub.io/ Nevertheless, I am not sure if this is the correct approach in your case and probably would recommend writing your own web API to talk to a database on your own machine.

**Reply by justinclift** [reply]
Hmmm, it'd probably work decently well from the API using "Live" SQLite databases.

That's what we call SQLite databases on DBHub.io that aren't just read-only.  They're fully read-write, so can have INSERT/UPDATE/DELETE run on them (etc) using the API.

There are various libraries for the DBHub.io API already around, including Go, Python, and someone's written one for NodeJS too (I'd have to go find it again).

**Reply by CacaceNicolas** [reply]
That is awesome! I tried set up a live database in DBHub, but there is no library to use the DBHub.io API in Godot, is there a way I can use the API in Godot?

**Reply by justinclift** [reply]
That's a good question.  Does the Godot library support making HTTP calls, and sending/receiving JSON over HTTP?

Because even if there's no existing library, if you can do HTTP calls you can manually create some library functions yourself to make it work.

For example, with the `Query()` function:

* https://api.dbhub.io/#query

All you really need to do for that is call https://api.dbhub.io/v1/query with some parameters to tell it what you're wanting to do.

Does that sound workable? :smile:

**Reply by CacaceNicolas** [reply]
Yes it does. I almost got it, but i keep getting the "Invalid database name" error.
This is the url i am using; https://api.dbhub.io/v1/columns?apikey=snipped&dbowner=Kony47&dbname=data

It seems to be the right way to send parameters because the server accepts my apikey. Is there something wrong with the way i am sending the dbname parameter?

**Reply by justinclift** [reply]
Hmmm, I've just edited your post to remove the api key as it *looked* like it could be a valid one. (yeah, I didn't log in to double check) :wink:

Looking at the database name, it should probably be `data.db` to match the actual file name.  With that done, it should (in theory) work. :smile:

**Reply by CacaceNicolas** [reply]
don't worry, it wasn't a real one (thanks though 😉), i tried that and many other variations, but nothing seems to work! any other idea of what could be wrong?
(thanks for your patience 😄)

**Reply by justinclift** [reply]
Oh.  It might need the `.` character encoded.

Actually, gimme a few minutes.  I'll go and load our server code in my IDE and refresh my memory of what it needs. :smile:

**Reply by justinclift** [reply]
Oh.  It's because the server is looking for HTTP `POST` requests, not HTTP `GET` requests.

So it's ignoring all of the parameter info that's appended to the url.

Hopefully that's an easy change to make on your end? :smile:

**Reply by justinclift** [reply]
Looking at the Godot docs, it seems like POST requests shouldn't be too hard.

https://docs.godotengine.org/en/stable/tutorials/networking/http_request_class.html#sending-data-to-the-server

It looks like the `$HTTPRequest.request()` call can take a parameter telling it the type of request to make, so setting that to `HTTPClient.METHOD_POST` should do the trick.  In theory. :smile:

```
var url = "https://api.dbhub.io/v1/tables"
var json = JSON.stringify(data_to_send)
var headers = ["Content-Type: application/json"]
$HTTPRequest.request(url, headers, HTTPClient.METHOD_POST, json)
```

**Reply by CacaceNicolas** [reply]
It still doesn't work 😔 but, i can tell it's not ignoring all of the parameters because if i change the api key, the error msg changes to: "Incorrect or unknown API key and certificate". Do you know anything else that could be wrong?

**Reply by justinclift** [reply]
Is it still complaining about the database name?

Hmmm, if you dump the JSON string (that you're sending) to the server to your screen instead, would you be ok to paste it here?  Minus anything confidential of course.

I'll try with the exact same string but watch what's happening from the server side.  It's likely to be something simple, that once we figure it out will open the doors for you to everything else. :smile:

**Reply by CacaceNicolas** [reply]
I am not trying to send a file, first i wanted to test it out with a reading request, so i was trying to use the "tables" function. does that change anything?

**Reply by justinclift** [reply]
> first i wanted to test it out with a reading request, so i was trying to use the "tables" function.

That's a good approach.  "Start simple first and get it working (at all).  Fancy stuff can come later."

I'll take a look at the server logs and see if it's recording enough info to tell what it's unhappy with.  In theory the info should be there somewhere. :smile:

**Reply by justinclift** [reply]
Ahhh, I can see the problem.  The HTTP calls you're doing are still all GET requests rather than POST requests.  From the server side logs (with sensitive info redacted):

```
111.222.333.444:52532 - Kony47 [2023-12-24T04:31:52.083486353Z] "GET /v1/tables?apikey=snipped&dbname=data.db&dbowner=Kony47 HTTP/1.1" "-" "-" "" "GodotEngine/3.5.2.stable.official (Windows)"
111.222.333.444:52845 - Kony47 [2023-12-24T04:43:44.753943134Z] "GET /v1/upload?apikey=snipped&dbname=data.db&dbowner=Kony47 HTTP/1.1" "-" "-" "" "GodotEngine/3.5.2.stable.official (Windows)"
```

It sounds like that call to `$HTTPRequest.request()` needs a bit of adjustment still. :smile:

**Reply by justinclift** [reply]
As a thought for helping you debug this, are you ok with running stuff from the command line (aka powershell on windows)?

There's a command line utility called `curl` which is commonly used for testing out HTTP servers, and the API documentation includes the `curl` command equivalent:

* https://api.dbhub.io/#tables

```
curl -F apikey="YOUR_API_KEY_HERE" -F dbowner="Kony47" -F dbname="data.db" https://api.dbhub.io/v1/tables
```

**Reply by CacaceNicolas** [reply]
The request works fine if I do it from the command line, so I guess the error is in the way I'm sending the request from Godot. This is what it looks like:

$HTTPRequest.request(url, [], HTTPClient.METHOD_POST)

The url I'm usign is: https://api.dbhub.io/v1/tables?apikey=snipped&dbowner=Kony47&dbname=data

Can you spot anything? Maybe I need some custom headers?

**Reply by justinclift** [reply]
> Can you spot anything?

Yep.  I *think* it'll be the missing "json" parameter on the end.  So instead it should be something like this:

```
$HTTPRequest.request(url, [], HTTPClient.METHOD_POST, json)
```

The json is where you put the parameter data.  Things like the database owner and name, the sql query for the functions that need one, and so on.

I'm not personally sure how to construct json data with Godot, but it's probably fairly straight forward to get started with.  Maybe some kind of list of key:value pairs or similar.

---

Oh, it looks fairly straight forward:

https://gamedevacademy.org/gdscript-json-tutorial-complete-guide/

And the "sending data to a server" example in the Godot docs looks to build upon that.  It has code showing how to send JSON to the server:

https://docs.godotengine.org/en/stable/tutorials/networking/http_request_class.html#sending-data-to-the-server

```
var json = JSON.stringify(data_to_send)
var headers = ["Content-Type: application/json"]
$HTTPRequest.request(url, headers, HTTPClient.METHOD_POST, json)
```

The `JSON.stringify()` there is turning some kind of structure (maybe a dictionary) into a JSON string, and the `$HTTPRequest.request()` is sending that string to the server.  That's what you want. :smile:

**Reply by justinclift** [reply]
As a data point, I've just logged into the server and taken a look at its record of your last few request.  They're correctly coming through as POST type requests now.  Like this:

```
111.222.333.444:55568 - Kony47 [2023-12-24T19:44:25.473423845Z] "POST /v1/execute?dbname=datadb&apikey=snipped&dbowner=Kony47&sql=snippedtoo HTTP/1.1" "-" "-" "" "GodotEngine/3.5.2.stable.official (Windows)"
```

The `POST` thing there is what you want.  The `?dbname=...` stuff on the end though isn't right, as that structure for data is only suitable for GET requests.

Instead of the parameters being added to the end of the url like that, they get put into a JSON structure and attached to the body of the request.

In theory, that's what the `$HTTPRequest.request(url, headers, HTTPClient.METHOD_POST, json)` call in your GDScript should do.  Hopefully. :smile:

**Reply by CacaceNicolas** [reply]
It still doesn't work. 😔 

![Captura de pantalla 2023-12-25 015217](https://github.com/sqlitebrowser/sqlitebrowser/assets/147722690/0690e8b6-c45c-4c91-92c5-82e098d26144)

That's my hole code, i used the "to_json()" method instead of "JSON.stringify()" because it doesn't exist in the version of godot I'm using. I also put the apikey in the url, because if I put it in the json file, the server doesn't read it properly. 🤷‍♂️

I feel really confused. Is there some information at the server end that might help? because if not, I think we've reached a dead end. 😕

**Reply by CacaceNicolas** [reply]
ok, I just did two identical requests, one from Godot, and one from the windows cmd, usign curl. Those are my two most recent requests. Can you see any difference between them in the server log?

**Reply by justinclift** [reply]
Hang on, I'll take a look. :smile:

**Reply by justinclift** [reply]
These are your last two requests, as shown in the server logs:

```
111.222.333.444:65381 - Kony47 [2023-12-26T02:37:27.136478318Z] "POST /v1/execute?apikey=snipped HTTP/1.1" "-" "-" "" "GodotEngine/3.5.2.stable.official (Windows)"
111.222.333.444:65385 - Kony47 [2023-12-26T02:37:49.651920901Z] "POST /v1/execute HTTP/1.1" "-" "-" "" "curl/8.4.0"
```

Try putting the api key in the json as well. :smile:

Note that at the moment the server is only logging the url of the request (as above) without capturing any of the POST data.

If putting the api key into the json data *doesn't work*, I can adjust the server so it captures the POST data for your requests to a local file for manual inspection.  That should let us get to the bottom of this. :smile:

**Reply by CacaceNicolas** [reply]
I put the apikey in the json and the error msg was "Incorrect or unknown API key and certificate" instead of "invalid database name", i guess this is because the server is completely ignoring the body of my request and the parameters in it.

**Reply by justinclift** [reply]
I'll adjust the server code to capture the POST data of your requests to a local file so we can figure out what's going wrong.

Sorry it's taking so long to get this worked out.  Like a lot of things though, persistence is king. :smile:

**Reply by justinclift** [reply]
Still working on adding that server code to capture the POST data, it's turning out to be a bit trickier than I'd like...

**Reply by justinclift** [reply]
Arghh.  I've figured out what's going wrong.

Although the server returns JSON for it's responses, the incoming request actually needs to be structured as multipart form-encoded data rather than JSON. :facepalm: :facepalm: :facepalm:

(We should probably change that for "version 2" of the API, but that'll be a task for some other time)

When doing a dump of the (working) curl POST data just now on my local dev server, it was showing this:

```
--------------------------I8a9SCP3px4qbjiKO8RoE5
Content-Disposition: form-data; name="apikey"

snipped
--------------------------I8a9SCP3px4qbjiKO8RoE5
Content-Disposition: form-data; name="dbowner"

default
--------------------------I8a9SCP3px4qbjiKO8RoE5
Content-Disposition: form-data; name="dbname"

data.db
--------------------------I8a9SCP3px4qbjiKO8RoE5
Content-Disposition: form-data; name="sql"

RGVsZXRlIGZyb20gQ2FydGFzUG9ySnVnYWRvcg==
--------------------------I8a9SCP3px4qbjiKO8RoE5--
```

That explains why it was ignoring everything you were sending it. :frowning:

Really sorry about giving you incorrect pointers back there.  I could have sworn we accepted JSON input as well, but clearly not (yet).

**Reply by justinclift** [reply]
This seems to be a successful example of sending multipart form data using Godot 3.x:

https://forum.godotengine.org/t/uploading-file-to-server-through-rest/6402/2

**Reply by CacaceNicolas** [reply]
![Captura de pantalla 2023-12-26 133245](https://github.com/sqlitebrowser/sqlitebrowser/assets/147722690/05ff2962-3e20-4dc1-93ad-52aa05a06da5)

Can you tell whats wrong? because it is still not working. I think that sending the body as JSON is wrong, but if i leave it as a PoolByteArray, the engine doesn't let me send the request. I tried sending it as a string too, but the same happened. Maybe there is some useful info at the server end. thanks for all your help. 😄

**Reply by justinclift** [reply]
Cool, I think we're on the right track.

I'll get that code which captures POST data to a local file finished and deployed on the server so we can tell what's actually going wrong, then figure out how to fix it.

It could take me a few hours, as I've just woken up after getting some sleep and need to get food/coffee/check-email-for-anything-super-urgent etc. :smile:

**Reply by justinclift** [reply, marked as answer]
Oh, in your `$HTTPRequest.request()` call, is there a way to convert `body` to a string directly rather than use `to_json()`?

---

Looking at the Godot 3.x docs, it looks like there's a function called `get_string_from_utf8()` which might be what's needed instead:

https://docs.godotengine.org/en/3.0/classes/class_poolbytearray.html#class-poolbytearray-get-string-from-utf8

So, that'd make it something like this:

```
$HTTPRequest.request(url, headers, true, HTTPClient.METHOD_POST, body.get_string_from_utf8())
```

**Reply by justinclift** [reply]
Ok, the `Execute` end point (and just that one) on the API server is now logging the POST data for your requests.  So we should be able to tell what's going wrong from here on.

Let me know when you've tried the execute end point so I can take a look. :smile:

**Reply by CacaceNicolas** [reply]
IT WORKED!!! FINALLY!!!

Thank you so much! you've been really helpful! 😄😄😄

**Reply by justinclift** [reply]
Sweet!  I guess that means I can remove the code that captures the POST data then. :smile:

As a thought, it *might* be worth trying to use a `PoolStringArray` instead of a `PoolByteArray`:

https://docs.godotengine.org/en/3.0/classes/class_poolstringarray.html

The example that I pointed to earlier was probably using a byte array because they wanted to upload a file or something.  In this case though, you're mostly just working with string data.

Oh... with the exception that for the `Upload` API end point you'd probably still want to use a ByteArray.

Hmmm, maybe ignore that as I might just be overcomplicating things. :smile:

**Reply by CacaceNicolas** [reply]
Thanks for the advice! I'll give it a shot

**Comment by justinclift** [comment]
@SanmiCace Hmmm, are you seeing errors on your end when trying to retrieve the table list for your databases?

I'm seeing what looks like an error message being spat out on the back end, which looks like something caused by a uppercase vs lowercase issue (a problem on our end, not yours).

But I'm not sure if that error is making its way to you or not.  ?

**Reply by justinclift** [reply]
Hmmm, I'm going to track down that error regardless.  I don't like seeing errors in our backend logs, and it could mean something isn't quite right...

**Reply by justinclift** [reply]
Ahhh, found the cause.  It's just a side effect of capitalisation stuff, so won't have any real effect.  All good. :smile:
